Pre: import necessary packages & import datasets 

In [1]:
import os
import pandas as pd
import numpy as np
import fancyimpute as fancy 


In [2]:
import warnings
warnings.filterwarnings("ignore")
# warnings.filterwarnings("ignore", category=DeprecationWarning)

In [3]:
# import datasets
movie_info_dataset = pd.read_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Original Data/The Movies Dataset/movies_metadata.csv')
credits_dataset = pd.read_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Original Data/The Movies Dataset/credits.csv')
keywords_dataset = pd.read_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Original Data/The Movies Dataset/keywords.csv')
ratings = pd.read_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Original Data/The Movies Dataset/ratings_small.csv')


1. Examine datasets

In [4]:
#examining the shapes of datasets 

credits_dataset.shape
keywords_dataset.shape
movie_info_dataset.shape


(45466, 24)

In [5]:
#examine the content of movie: info, credits, and keywords

movie_info_dataset.info()
keywords_dataset.head()
credits_dataset.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

2. Clean movie_info set by dropping entires with no title & purging duplicates

In [6]:
# view titless entries
movie_info_dataset[movie_info_dataset['title'].isnull()]

#drop them
movie_info_dataset.dropna(subset = ['title'], inplace=True)

3. Convert movie 'id' column type to 'int' in order to join datasets

In [7]:
#convert 'id' to 'int' datatype

movie_info_dataset['id']  = movie_info_dataset['id'].astype(int)

In [8]:
# join movie_info with credits dataframe. call new df 'movie_stats'

movie_stats = movie_info_dataset.merge(credits_dataset, on='id')
movie_stats.head(2)


adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                               homepage    id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
1                                   NaN  8844  tt0113497                en   

  original_title                                           overview  ...  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1        Jumanji  When siblings Judy and Peter discover an encha...  ...   

  runtime                                   spoken_languages    status  \
0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   

                                     tagline      title  video  vote_average  \
0                                        NaN  Toy Story  False           7.7   
1  Roll the dice and unleash the excitement!    Jumanji  False           6.9   

  vote_count                                               cast  \
0     5415.0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1     2413.0  [{'cast_id': 1, 'character': 'Alan Parrish', '...   

                                                crew  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...  

[2 rows x 26 columns]

In [9]:
#join movie_stats with 'keywords' dataframe

movie_stats = movie_stats.merge(keywords_dataset, on='id')
movie_stats.head(2)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

                               homepage    id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
1                                   NaN  8844  tt0113497                en   

  original_title                                           overview  ...  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1        Jumanji  When siblings Judy and Peter discover an encha...  ...   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   

                                     tagline      title  video  vote_average  \
0                                        NaN  Toy Story  False           7.7   
1  Roll the dice and unleash the excitement!    Jumanji  False           6.9   

   vote_count                                               cast  \
0      5415.0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1      2413.0  [{'cast_id': 1, 'character': 'Alan Parrish', '...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   

                                            keywords  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...  
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...  

[2 rows x 27 columns]

In [10]:
# find duplicate entires based on movie id 

show_dups = movie_stats[movie_stats.duplicated(subset='id',keep='first') == True]
show_dups.head(2)

adult belongs_to_collection budget  \
677  False                   NaN      0   
678  False                   NaN      0   

                                                genres homepage      id  \
677  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...      NaN  105045   
678  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...      NaN  105045   

       imdb_id original_language   original_title  \
677  tt0111613                de  Das Versprechen   
678  tt0111613                de  Das Versprechen   

                                              overview  ...  \
677  East-Berlin, 1961, shortly after the erection ...  ...   
678  East-Berlin, 1961, shortly after the erection ...  ...   

                             spoken_languages    status  \
677  [{'iso_639_1': 'de', 'name': 'Deutsch'}]  Released   
678  [{'iso_639_1': 'de', 'name': 'Deutsch'}]  Released   

                     tagline        title  video  vote_average  vote_count  \
677  A love, a hope, a wall.  The Promise  False           5.0         1.0   
678  A love, a hope, a wall.  The Promise  False           5.0         1.0   

                                                  cast  \
677  [{'cast_id': 5, 'character': 'Sophie II', 'cre...   
678  [{'cast_id': 5, 'character': 'Sophie II', 'cre...   

                                                  crew  \
677  [{'credit_id': '52fe4a44c3a36847f81c463f', 'de...   
678  [{'credit_id': '52fe4a44c3a36847f81c463f', 'de...   

                                              keywords  
677  [{'id': 7059, 'name': 'anti-communism'}, {'id'...  
678  [{'id': 7059, 'name': 'anti-communism'}, {'id'...  

[2 rows x 27 columns]

In [11]:
# lets drop any entries that share 'ids' (duplicates)

movie_stats.drop_duplicates(subset='id', keep='first', inplace=True)

Now we have the 3 movie information data sets merged into one.

4. Explore and drop useless columns

In [12]:
# viewing all columns 
movie_stats.columns

Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count', 'cast', 'crew', 'keywords'],
      dtype='object')

In [13]:
#see which columns contain missing values
movie_stats.isna().sum()

adult                        0
belongs_to_collection    40942
budget                       0
genres                       0
homepage                 37656
id                           0
imdb_id                     17
original_language           11
original_title               0
overview                   954
popularity                   0
poster_path                383
production_companies         0
production_countries         0
release_date                84
revenue                      0
runtime                    257
spoken_languages             0
status                      81
tagline                  25028
title                        0
video                        0
vote_average                 0
vote_count                   0
cast                         0
crew                         0
keywords                     0
dtype: int64

Looks like 'belongs_to_collection' 'homepage' and 'tagline' have lots of missing values and aren't necessary.

Lets explore some other columns that seem unimportant.

In [14]:
# examining the 'adult' column
movie_stats['adult'].value_counts() # only 9 with True value
movie_stats[movie_stats['adult'] == 'True'] # mostly sex related films with low popularity scores and low 'vote_counts'

adult belongs_to_collection  budget  \
19679  True                   NaN       0   
28914  True                   NaN       0   
32142  True                   NaN       0   
32321  True                   NaN       0   
41065  True                   NaN       0   
41066  True                   NaN       0   
41736  True                   NaN       0   
42171  True                   NaN  750000   
44251  True                   NaN       0   

                                                  genres  \
19679                     [{'id': 27, 'name': 'Horror'}]   
28914                      [{'id': 80, 'name': 'Crime'}]   
32142                     [{'id': 35, 'name': 'Comedy'}]   
32321  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
41065  [{'id': 80, 'name': 'Crime'}, {'id': 27, 'name...   
41066  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...   
41736  [{'id': 80, 'name': 'Crime'}, {'id': 27, 'name...   
42171  [{'id': 53, 'name': 'Thriller'}, {'id': 80, 'n...   
44251  [{'id': 878, 'name': 'Science Fiction'}, {'id'...   

                                   homepage      id    imdb_id  \
19679                                   NaN    5422  tt0079642   
28914  http://www.imdb.com/title/tt0119931/  169158  tt0119931   
32142                                   NaN   44781  tt0322232   
32321             http://www.dietofsex.com/  324230  tt3094816   
41065                                   NaN   35731  tt1161951   
41066                                   NaN   55774  tt1153101   
41736                                   NaN   39922  tt0070696   
42171                                   NaN  376004  tt4044464   
44251                                   NaN  448290  tt2697688   

      original_language                       original_title  \
19679                it  Le notti erotiche dei morti viventi   
28914                en                             Standoff   
32142                cn                                發電悄嬌娃   
32321                es                          Diet of Sex   
41065                en           Amateur Porn Star Killer 2   
41066                en                             The Band   
41736                da                              Dværgen   
42171                en                           Adulterers   
44251                en                           Half -Life   

                                                overview  ...  \
19679  A sailor takes an American businessman and his...  ...   
28914  The FBI and Texas police make ready to storm t...  ...   
32142  Electrical Girl centers around a horny young w...  ...   
32321  Ágata suffers from a psychological disorder wh...  ...   
41065  Shane Ryan's sequel to the disturbing Amateur ...  ...   
41066  Australian film about a fictional sub-par Aust...  ...   
41736  Olaf and his mother run a boarding house and a...  ...   
42171  A man who returns home to find his wife cheati...  ...   
44251  An American scientific research facility opens...  ...   

                                 spoken_languages         status  \
19679   [{'iso_639_1': 'it', 'name': 'Italiano'}]       Released   
28914    [{'iso_639_1': 'en', 'name': 'English'}]       Released   
32142  [{'iso_639_1': 'cn', 'name': '广州话 / 廣州話'}]       Released   
32321    [{'iso_639_1': 'es', 'name': 'Español'}]       Released   
41065    [{'iso_639_1': 'en', 'name': 'English'}]       Released   
41066    [{'iso_639_1': 'en', 'name': 'English'}]       Released   
41736    [{'iso_639_1': 'en', 'name': 'English'}]       Released   
42171    [{'iso_639_1': 'en', 'name': 'English'}]       Released   
44251    [{'iso_639_1': 'en', 'name': 'English'}]  In Production   

                                              tagline  \
19679                                             NaN   
28914                       Trust Will Get You Killed   
32142                                             NaN   
32321       Comedy, food, drama and sex, a lot of sex   
41065  Humiliation. Rape. Murder. Y

Looks like the 'adult' column can be dropped.

In [15]:
#checking out the 'status' column to see how many films are unreleased

movie_stats['status'].value_counts() 

Released           44984
Rumored              229
Post Production       98
In Production         20
Planned               15
Canceled               2
Name: status, dtype: int64

Around 350 movies on the list haven't been released, seeing as we cannot recommend an unreleased film, we will drop them and also drop the 'staus' column.

In [16]:
#drop entries that have not been released
unreleased = movie_stats[movie_stats['status'] != 'Released'].index
movie_stats.drop(unreleased, inplace=True)

In [17]:
# examine final suspicious columns
movie_stats['video'].value_counts()  # only 93 films have a 'True' value for video. Doesn't seem like a useful metric  
movie_stats['poster_path'].value_counts() #looks like a link to  film poster. Won't be relevant.
movie_stats['original_title'] # mostly similar to the final and known title. Won't be relevant.
movie_stats['imdb_id'] # Anothe unique ID. Doesn't seem useful as we already have a unique ID for each film. 

0        tt0114709
1        tt0113497
2        tt0113228
3        tt0114885
4        tt0113041
           ...    
46619    tt6209470
46620    tt2028550
46621    tt0303758
46622    tt0008536
46623    tt6980792
Name: imdb_id, Length: 44984, dtype: object

There are a total of 9 columns we have deemed unnecessary and thus will be dropped from our dataset.

In [18]:
# drop sunnecessary columns 
movie_stats.drop(['adult', 'homepage', 'status', 'video', 'poster_path', 'original_title', 'imdb_id', 'belongs_to_collection', 'tagline'], axis=1, inplace=True)

5. Lets examine and clean our features to be more legible 

In [19]:
# view our dataset
movie_stats.columns

Index(['budget', 'genres', 'id', 'original_language', 'overview', 'popularity',
       'production_companies', 'production_countries', 'release_date',
       'revenue', 'runtime', 'spoken_languages', 'title', 'vote_average',
       'vote_count', 'cast', 'crew', 'keywords'],
      dtype='object')

Rearrange the columns with more ordinality (title and id first, etc.)

In [20]:
movie_stats = movie_stats.reindex(['id','title','genres','keywords','cast', 'crew','overview', 'popularity','runtime','release_date','budget','revenue','original_language','spoken_languages', 'production_companies','production_countries','vote_average','vote_count'], axis=1)
movie_stats.head(3)

id             title                                             genres  \
0    862         Toy Story  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1   8844           Jumanji  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  15602  Grumpier Old Men  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   

                                            keywords  \
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...   
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...   
2  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...   

                                            overview popularity  runtime  \
0  Led by Woody, Andy's toys live happily in his ...  21.946943     81.0   
1  When siblings Judy and Peter discover an encha...  17.015539    104.0   
2  A family wedding reignites the ancient feud be...    11.7129    101.0   

  release_date    budget      revenue original_language  \
0   1995-10-30  30000000  373554033.0                en   
1   1995-12-15  65000000  262797249.0                en   
2   1995-12-22         0          0.0                en   

                                    spoken_languages  \
0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2           [{'iso_639_1': 'en', 'name': 'English'}]   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   

                                production_countries  vote_average  vote_count  
0  [{'iso_3166_1': 'US', 'name': 'United States o...           7.7      5415.0  
1  [{'iso_3166_1': 'US', 'name': 'United States o...           6.9      2413.0  
2  [{'iso_3166_1': 'US', 'name': 'United States o...           6.5        92.0

Some of our columns are dictionaries packed in to lists and need to be unpacked.

In [21]:
# cleaning some columns that contain hard to read descriptions 
import ast
def clean(hard):
    lst = []
    for i in ast.literal_eval(hard):
        lst.append(i['name'])
    return lst 

movie_stats['production_companies'] = movie_stats['production_companies'].apply(clean)
movie_stats['production_countries'] = movie_stats['production_countries'].apply(clean)
movie_stats['spoken_languages'] = movie_stats['spoken_languages'].apply(clean)
movie_stats['genres'] = movie_stats['genres'].apply(clean)
movie_stats['keywords'] = movie_stats['keywords'].apply(clean)
movie_stats['cast'] = movie_stats['cast'].apply(clean)

The 'crew' feature will list dozens of people so lets refine that column and select just the director.

In [22]:
# function for retrieving 'director' from crew list 
def get_director(obj):
    lst = []
    count = 0
    for i in ast.literal_eval(obj):
        if i['job'] == 'Director':
            lst.append(i['name'])
            break
    return lst

movie_stats['crew'] = movie_stats['crew'].apply(get_director)

In [23]:
#rename 'crew' column to 'driector'
 
movie_stats.rename(columns={'crew': 'director'}, inplace=True)

In [24]:
# viewing our cleaned 'movie stats' data frame 
movie_stats.dtypes


id                        int64
title                    object
genres                   object
keywords                 object
cast                     object
director                 object
overview                 object
popularity               object
runtime                 float64
release_date             object
budget                   object
revenue                 float64
original_language        object
spoken_languages         object
production_companies     object
production_countries     object
vote_average            float64
vote_count              float64
dtype: object

6. Time to convert some datatypes

Lets convert our 'popularity' and 'budget' columns to float datatypes, being that it is a number ranking. 

In [25]:
movie_stats['popularity'] = pd.to_numeric(movie_stats['popularity'])
movie_stats['budget'] = pd.to_numeric(movie_stats['budget'])


Our 'release_date' column can be refined to just 'year'.

In [26]:
#convert 'release_date' column to datetime datatype
movie_stats['release_date'] = pd.to_datetime(movie_stats['release_date'])

#extract year from date
movie_stats['year'] = movie_stats['release_date'].dt.year


In [27]:
# 78 films have missing  value for release year, enter 0 for those 
movie_stats.year.fillna(0, inplace=True)

# convert years to int with no decimal place 
movie_stats['year'] = movie_stats['year'].astype('int')

In [28]:
# retitle 'year' to 'release year'
movie_stats['release year'] = movie_stats['year']

In [29]:
# drop 'release_date' and 'year' columns 
movie_stats.drop(['release_date', 'year'], axis=1, inplace=True)


In [30]:
#convert some column dtypes from float to int
movie_stats['revenue'] = movie_stats['revenue'].astype('int')
movie_stats['vote_count'] = movie_stats['vote_count'].astype('int')

In [31]:
movie_stats.dtypes

id                        int64
title                    object
genres                   object
keywords                 object
cast                     object
director                 object
overview                 object
popularity              float64
runtime                 float64
budget                    int64
revenue                   int64
original_language        object
spoken_languages         object
production_companies     object
production_countries     object
vote_average            float64
vote_count                int64
release year              int64
dtype: object

Here we have a pretty cleaned up dataset. We are now going to make a couple copies of our dataset to prepare for different modeling approaches. 

In [32]:
movie_num = movie_stats.copy()
movie_text = movie_stats.copy()

First we are going to focus on a features that we can 'one-hot-encode' - trasforming categeories into numerical representations. 

Pt. 2 One-hot-encoding values

Lets take a look at the 'original_language' and 'spoken_langue' columns to see if we need both.

In [ ]:
movie_num['original_language'].value_counts().head(20) # ~32,000 films are originally in English
movie_num['spoken_languages'].value_counts().head(30) # ~22,000 are English only, and several thousand more include english as a language.  

In [34]:
type(movie_num.loc[0, 'production_countries'])

list

After doing some research, we discovered that any language spoken in the film can count as a 'spoken language'. What is more important for our sake is whether or not the movie is a foreign film. So we will drop the 'spoken_languages' column. 

1. Drop 'spoken languages' column

In [35]:
# drop 'spoken languages' column
movie_num.drop(['spoken_languages'], axis=1, inplace=True)

In [ ]:
# checking which movies have missing original lang value
movie_num[movie_num['original_language'].isna()]

In [37]:
# imputing missing language value
movie_num.loc[19764, 'original_language'] = 'de'
movie_num.loc[21800, 'original_language'] = 'en'
movie_num.loc[23028, 'original_language'] = 'en'
movie_num.loc[32349, 'original_language'] = 'en'
movie_num.loc[38572, 'original_language'] = 'cs'
movie_num.loc[42209, 'original_language'] = 'en'
movie_num.loc[43034, 'original_language'] = 'en'
movie_num.loc[45218, 'original_language'] = 'fr'
movie_num.loc[45571, 'original_language'] = 'en'
movie_num.loc[45737, 'original_language'] = 'de'

3. Make a copy of dataset to focus soley on features dealing with language and country 

In [38]:
# make a copy of our dataset 
movie_nati = movie_num[['id', 'title', 'original_language', 'production_countries']]
movie_nati

id                        title original_language  \
0         862                    Toy Story                en   
1        8844                      Jumanji                en   
2       15602             Grumpier Old Men                en   
3       31357            Waiting to Exhale                en   
4       11862  Father of the Bride Part II                en   
...       ...                          ...               ...   
46619  439050                       Subdue                fa   
46620  111109          Century of Birthing                tl   
46621   67758                     Betrayal                en   
46622  227506             Satan Triumphant                en   
46623  461257                     Queerama                en   

             production_countries  
0      [United States of America]  
1      [United States of America]  
2      [United States of America]  
3      [United States of America]  
4      [United States of America]  
...                           ...  
46619                      [Iran]  
46620               [Philippines]  
46621  [United States of America]  
46622                    [Russia]  
46623            [United Kingdom]  

[44984 rows x 4 columns]

Let's take a look at all the languages we will list as 'foreign', without any US bias (lol)

In [39]:
# list of foreign languages (sans english)
for_lan_list = list(movie_nati['original_language'].unique())
for_lan_list.remove('en')
for_lan_list[:10]

['fr', 'zh', 'it', 'fa', 'nl', 'de', 'cn', 'ar', 'es', 'ru']

4. Create new columns for 'english language' and 'foreign language'

In [40]:
#create new column that has '1' if original language is English, '0' if not.
movie_nati['english_lang'] = np.where(movie_nati['original_language'] == 'en', 1, 0)

#create new column that has '1' if original language is not English, '0' if it is.
movie_nati['foreign_lang'] = np.where(movie_nati['original_language'] != 'en', 1, 0)

The step above may be redundant, but now we have two new features; one for English language films and one for foreign language films.

In [41]:
# now we can drop 'original_language' column
movie_nati.drop(['original_language'], axis=1, inplace=True)

5. Now we'll look at the 'production_countries' column

In [42]:
# breaking out the 'production_countries' list and filter for films that have 'United States of America' in the list
unique_country = movie_nati.explode('production_countries')
uniq_country = unique_country[unique_country['production_countries'] == 'United States of America']
uniq_country

id                        title      production_countries  \
0         862                    Toy Story  United States of America   
1        8844                      Jumanji  United States of America   
2       15602             Grumpier Old Men  United States of America   
3       31357            Waiting to Exhale  United States of America   
4       11862  Father of the Bride Part II  United States of America   
...       ...                          ...                       ...   
46614   84419             House of Horrors  United States of America   
46616  289923          The Burkittsville 7  United States of America   
46617  222848              Caged Heat 3000  United States of America   
46618   30840                   Robin Hood  United States of America   
46621   67758                     Betrayal  United States of America   

       english_lang  foreign_lang  
0                 1             0  
1                 1             0  
2                 1             0  
3                 1             0  
4                 1             0  
...             ...           ...  
46614             1             0  
46616             1             0  
46617             1             0  
46618             1             0  
46621             1             0  

[21005 rows x 5 columns]

In [43]:
# create function to count instances of countries in 'production_countries' column

def to_1D(series):
 return pd.Series([x for _list in series for x in _list])

to_1D(movie_nati['production_countries']).value_counts().head()

United States of America    21005
United Kingdom               4074
France                       3893
Germany                      2235
Italy                        2144
dtype: int64

6. Now we will create 3 new columns; 'US_film' , 'foreign_film', 'international' (which means more than one country listen in 'production_countries')

In [44]:
# using lambda functions to break down 'production countrie' column 
movie_nati['US_film'] = movie_nati['production_countries'].apply(lambda x: 1 if x[0:1] == ['United States of America'] else 0)
movie_nati['foreign_film'] = movie_nati['production_countries'].apply(lambda x: 1 if x[0:1] != ['United States of America'] and len(x) == 1 else 0)
movie_nati['international'] = movie_nati['production_countries'].apply(lambda x: 1 if len(x) > 1 else 0)


7. Now lets drop the 'production countries' column 

In [45]:
movie_nati.drop(['production_countries'], axis=1, inplace=True)

In [46]:
# lets take a look at our dateframe
movie_nati.head(5)

id                        title  english_lang  foreign_lang  US_film  \
0    862                    Toy Story             1             0        1   
1   8844                      Jumanji             1             0        1   
2  15602             Grumpier Old Men             1             0        1   
3  31357            Waiting to Exhale             1             0        1   
4  11862  Father of the Bride Part II             1             0        1   

   foreign_film  international  
0             0              0  
1             0              0  
2             0              0  
3             0              0  
4             0              0

Looks good!

8. Now lets merge our newly created US/foreign dataframe with our movie_num dataframe

In [47]:
movie_nums = pd.merge(movie_num, movie_nati, how='inner', left_on=['id','title'], right_on=['id','title'])

In [48]:
movie_nums.drop(['original_language', 'production_countries'], axis=1, inplace=True)

In [49]:
movie_nums.head(2)

id      title                        genres  \
0   862  Toy Story   [Animation, Comedy, Family]   
1  8844    Jumanji  [Adventure, Fantasy, Family]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [board game, disappearance, based on children'...   

                                                cast         director  \
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  [John Lasseter]   
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst,...   [Joe Johnston]   

                                            overview  popularity  runtime  \
0  Led by Woody, Andy's toys live happily in his ...   21.946943     81.0   
1  When siblings Judy and Peter discover an encha...   17.015539    104.0   

     budget    revenue                               production_companies  \
0  30000000  373554033                          [Pixar Animation Studios]   
1  65000000  262797249  [TriStar Pictures, Teitler Film, Interscope Co...   

   vote_average  vote_count  release year  english_lang  foreign_lang  \
0           7.7        5415          1995             1             0   
1           6.9        2413          1995             1             0   

   US_film  foreign_film  international  
0        1             0              0  
1        1             0              0

Pt. 3 Let's one-hot-encode the 'production companies' feature!

1. Create new dataframe with only two features of interest

In [50]:
movie_comp = movie_nums[['id', 'title', 'production_companies']]

In [51]:
movie_comp

id                        title  \
0         862                    Toy Story   
1        8844                      Jumanji   
2       15602             Grumpier Old Men   
3       31357            Waiting to Exhale   
4       11862  Father of the Bride Part II   
...       ...                          ...   
44979  439050                       Subdue   
44980  111109          Century of Birthing   
44981   67758                     Betrayal   
44982  227506             Satan Triumphant   
44983  461257                     Queerama   

                                    production_companies  
0                              [Pixar Animation Studios]  
1      [TriStar Pictures, Teitler Film, Interscope Co...  
2                         [Warner Bros., Lancaster Gate]  
3               [Twentieth Century Fox Film Corporation]  
4           [Sandollar Productions, Touchstone Pictures]  
...                                                  ...  
44979                                                 []  
44980                                      [Sine Olivia]  
44981                          [American World Pictures]  
44982                                        [Yermoliev]  
44983                                                 []  

[44984 rows x 3 columns]

2. Lets grab the top 2 production companies for each film

In [52]:
#create two new columns for top 2 production companies 
movie_comp['prod_comp1'] = movie_comp['production_companies'].apply(lambda x: x[0:1])
movie_comp['prod_comp2'] = movie_comp['production_companies'].apply(lambda x: x[1:2])

In [53]:
# convert list to string
movie_comp['prod_comp1'] = movie_comp['prod_comp1'].str.get(0)
movie_comp['prod_comp2'] = movie_comp['prod_comp2'].str.get(0)

In [54]:
#lets check out the top most listed production companies 
prod_comps = movie_comp['production_companies'].explode()
prod_comps.value_counts().head(10)

Warner Bros.                              1248
Metro-Goldwyn-Mayer (MGM)                 1073
Paramount Pictures                         996
Twentieth Century Fox Film Corporation     834
Universal Pictures                         827
Columbia Pictures Corporation              447
Canal+                                     433
Columbia Pictures                          429
RKO Radio Pictures                         290
United Artists                             276
Name: production_companies, dtype: int64

3. Now we create feature columns for the top 10(ish) production companies 

In [55]:
# create function to find values in prod_comp1 and prod_comp2 columns
def prod_comp(row, studio, altname):
    if row['prod_comp1'] == studio or row['prod_comp2'] == studio: return 1
    elif row['prod_comp1'] == altname or row['prod_comp2'] == altname: return 1
    else: return 0


In [56]:
# envoking above function to create 14 new columns for top production companies and assign 1 or 0 if movie was made by said company

movie_comp['Paramount Pictures'] = movie_comp.apply(lambda x: prod_comp(x,'Paramount Pictures','x'), axis=1)
movie_comp['MGM'] = movie_comp.apply(lambda x: prod_comp(x,'Metro-Goldwyn-Mayer (MGM)','x'), axis=1)
movie_comp['20th Century Fox'] = movie_comp.apply(lambda x: prod_comp(x,'Twentieth Century Fox Film Corporation', 'Fox Searchlight Pictures'), axis=1)
movie_comp['Warner Bros.'] = movie_comp.apply(lambda x: prod_comp(x,'Warner Bros.','x'), axis=1)
movie_comp['Universal Pictures'] =movie_comp.apply(lambda x: prod_comp(x,'Universal Pictures','x'), axis=1)
movie_comp['Columbia Pictures'] = movie_comp.apply(lambda x: prod_comp(x,'Columbia Pictures', 'Columbia Pictures Corporation'), axis=1)
movie_comp['Disney'] = movie_comp.apply(lambda x: prod_comp(x,'Walt Disney Pictures', 'Walt Disney Productions'), axis=1)
movie_comp['Pixar'] = movie_comp.apply(lambda x: prod_comp(x,'Pixar Animation Studios','x'), axis=1)
movie_comp['New Line Cinema'] = movie_comp.apply(lambda x: prod_comp(x,'New Line Cinema','x'), axis=1)
movie_comp['United Artists'] = movie_comp.apply(lambda x: prod_comp(x,'United Artists','x'), axis=1)
movie_comp['Canal+'] = movie_comp.apply(lambda x: prod_comp(x,'Canal+', 'StudioCanal'), axis=1)
movie_comp['RKO Radio Pictures'] = movie_comp.apply(lambda x: prod_comp(x,'RKO Radio Pictures', 'x'), axis=1)
movie_comp['Touchstone Pictures'] = movie_comp.apply(lambda x: prod_comp(x,'Touchstone Pictures', 'x'), axis=1)
movie_comp['BBC Films'] = movie_comp.apply(lambda x: prod_comp(x,'BBC Films','British Broadcasting Corporation (BBC)'), axis=1)

4. Now lets drop the uneeded columns relating to production companies

In [57]:
# drop unneeded columns
movie_comp.drop(['production_companies', 'prod_comp1', 'prod_comp2'], axis=1, inplace=True)

In [58]:
movie_comp

id                        title  Paramount Pictures  MGM  \
0         862                    Toy Story                   0    0   
1        8844                      Jumanji                   0    0   
2       15602             Grumpier Old Men                   0    0   
3       31357            Waiting to Exhale                   0    0   
4       11862  Father of the Bride Part II                   0    0   
...       ...                          ...                 ...  ...   
44979  439050                       Subdue                   0    0   
44980  111109          Century of Birthing                   0    0   
44981   67758                     Betrayal                   0    0   
44982  227506             Satan Triumphant                   0    0   
44983  461257                     Queerama                   0    0   

       20th Century Fox  Warner Bros.  Universal Pictures  Columbia Pictures  \
0                     0             0                   0                  0   
1                     0             0                   0                  0   
2                     0             1                   0                  0   
3                     1             0                   0                  0   
4                     0             0                   0                  0   
...                 ...           ...                 ...                ...   
44979                 0             0                   0                  0   
44980                 0             0                   0                  0   
44981                 0             0                   0                  0   
44982                 0             0                   0                  0   
44983                 0             0                   0                  0   

       Disney  Pixar  New Line Cinema  United Artists  Canal+  \
0           0      1                0               0       0   
1           0      0                0               0       0   
2           0      0                0               0       0   
3           0      0                0               0       0   
4           0      0                0               0       0   
...       ...    ...              ...             ...     ...   
44979       0      0                0               0       0   
44980       0      0                0               0       0   
44981       0      0                0               0       0   
44982       0      0                0               0       0   
44983       0      0                0               0       0   

       RKO Radio Pictures  Touchstone Pictures  BBC Films  
0                       0                    0          0  
1                       0                    0          0  
2                       0                    0          0  
3                       0                    0          0  
4                       0                    1          0  
...                   ...                  ...        ...  
44979                   0                    0          0  
44980                   0                    0          0  
44981                   0                    0          0  
44982                   0                    0          0  
44983                   0                    0          0  

[44984 rows x 16 columns]

In [59]:
movie_nums = pd.merge(movie_nums, movie_comp, how='inner', left_on=['id','title'], right_on=['id','title'])

5. Looking good! Time to explore the genres column

Pt. 4 Encode the 'genres' feature

In [60]:
#make dataset with just genres
movie_genres = movie_nums[['id', 'title', 'genres']]

In [61]:
# explore the amount of genres 
genres = movie_genres['genres'].explode()
genres.value_counts()

Drama              20058
Comedy             13073
Thriller            7566
Romance             6663
Action              6542
Horror              4634
Crime               4283
Documentary         3872
Adventure           3468
Science Fiction     3013
Family              2740
Mystery             2453
Fantasy             2290
Animation           1912
Music               1587
Foreign             1586
History             1391
War                 1317
Western             1038
TV Movie             758
Name: genres, dtype: int64

There are 20 different genres.

6. Time to one-hot-encode the genres 

In [62]:
from sklearn.preprocessing import MultiLabelBinarizer

7. Implement our sklearn function to one-hot-encode our 'genres' column

In [63]:
mlb = MultiLabelBinarizer(sparse_output=True)

movie_genres = movie_genres.join(
            pd.DataFrame.sparse.from_spmatrix(
                mlb.fit_transform(movie_genres.pop('genres')),
                index=movie_genres.index,
                columns=mlb.classes_))

In [64]:
movie_genres

id                        title  Action  Adventure  Animation  \
0         862                    Toy Story       0          0          1   
1        8844                      Jumanji       0          1          0   
2       15602             Grumpier Old Men       0          0          0   
3       31357            Waiting to Exhale       0          0          0   
4       11862  Father of the Bride Part II       0          0          0   
...       ...                          ...     ...        ...        ...   
44979  439050                       Subdue       0          0          0   
44980  111109          Century of Birthing       0          0          0   
44981   67758                     Betrayal       1          0          0   
44982  227506             Satan Triumphant       0          0          0   
44983  461257                     Queerama       0          0          0   

       Comedy  Crime  Documentary  Drama  Family  ...  History  Horror  Music  \
0           1      0            0      0       1  ...        0       0      0   
1           0      0            0      0       1  ...        0       0      0   
2           1      0            0      0       0  ...        0       0      0   
3           1      0            0      1       0  ...        0       0      0   
4           1      0            0      0       0  ...        0       0      0   
...       ...    ...          ...    ...     ...  ...      ...     ...    ...   
44979       0      0            0      1       1  ...        0       0      0   
44980       0      0            0      1       0  ...        0       0      0   
44981       0      0            0      1       0  ...        0       0      0   
44982       0      0            0      0       0  ...        0       0      0   
44983       0      0            0      0       0  ...        0       0      0   

       Mystery  Romance  Science Fiction  TV Movie  Thriller  War  Western  
0            0        0                0         0         0    0        0  
1            0        0                0         0         0    0        0  
2            0        1                0         0         0    0        0  
3            0        1                0         0         0    0        0  
4            0        0                0         0         0    0        0  
...        ...      ...              ...       ...       ...  ...      ...  
44979        0        0                0         0         0    0        0  
44980        0        0                0         0         0    0        0  
44981        0        0                0         0         1    0        0  
44982        0        0                0         0         0    0        0  
44983        0        0                0         0         0    0        0  

[44984 rows x 22 columns]

Pt. 5 Dealing with 'director' and 'cast' 

In [65]:
# make new column with 'cast' and 'director'

movie_folks = movie_nums[['id','title', 'cast', 'director']]

In [66]:
# create new column with top 3 actors from ech film 

movie_folks['leads'] = movie_folks['cast'].apply(lambda x: x[:3])

In [67]:
movie_folks

id                        title  \
0         862                    Toy Story   
1        8844                      Jumanji   
2       15602             Grumpier Old Men   
3       31357            Waiting to Exhale   
4       11862  Father of the Bride Part II   
...       ...                          ...   
44979  439050                       Subdue   
44980  111109          Century of Birthing   
44981   67758                     Betrayal   
44982  227506             Satan Triumphant   
44983  461257                     Queerama   

                                                    cast            director  \
0      [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...     [John Lasseter]   
1      [Robin Williams, Jonathan Hyde, Kirsten Dunst,...      [Joe Johnston]   
2      [Walter Matthau, Jack Lemmon, Ann-Margret, Sop...     [Howard Deutch]   
3      [Whitney Houston, Angela Bassett, Loretta Devi...   [Forest Whitaker]   
4      [Steve Martin, Diane Keaton, Martin Short, Kim...     [Charles Shyer]   
...                                                  ...                 ...   
44979        [Leila Hatami, Kourosh Tahami, Elham Korda]  [Hamid Nematollah]   
44980  [Angel Aquino, Perry Dizon, Hazel Orencio, Joe...          [Lav Diaz]   
44981  [Erika Eleniak, Adam Baldwin, Julie du Page, J...    [Mark L. Lester]   
44982  [Iwan Mosschuchin, Nathalie Lissenko, Pavel Pa...  [Yakov Protazanov]   
44983                                                 []     [Daisy Asquith]   

                                                   leads  
0                    [Tom Hanks, Tim Allen, Don Rickles]  
1         [Robin Williams, Jonathan Hyde, Kirsten Dunst]  
2             [Walter Matthau, Jack Lemmon, Ann-Margret]  
3      [Whitney Houston, Angela Bassett, Loretta Devine]  
4             [Steve Martin, Diane Keaton, Martin Short]  
...                                                  ...  
44979        [Leila Hatami, Kourosh Tahami, Elham Korda]  
44980         [Angel Aquino, Perry Dizon, Hazel Orencio]  
44981       [Erika Eleniak, Adam Baldwin, Julie du Page]  
44982  [Iwan Mosschuchin, Nathalie Lissenko, Pavel Pa...  
44983                                                 []  

[44984 rows x 5 columns]

We don't need more than the top actors for each film so let's drop the cast column, keep just the 'leads' column 

In [68]:
# drop 'cast' column

movie_folks.drop(['cast'],axis=1,inplace=True)

In [69]:
# examine 'leads' counts 

to_1D(movie_folks['leads']).value_counts().head(11)

John Wayne           106
Jackie Chan           89
Michael Caine         86
Robert De Niro        84
Gérard Depardieu      82
Christopher Lee       74
Nicolas Cage          71
Bette Davis           68
James Mason           67
Samuel L. Jackson     67
Bruce Willis          64
dtype: int64

In [70]:
# examine 'director' counts  

to_1D(movie_folks['director']).value_counts().head(10)

John Ford           66
Michael Curtiz      65
Werner Herzog       54
Alfred Hitchcock    53
Georges Méliès      49
Woody Allen         49
Sidney Lumet        46
Jean-Luc Godard     46
Charlie Chaplin     44
Raoul Walsh         43
dtype: int64

In [71]:
# create copy of df to break down 'director' and 'leads' columns
one_df = movie_folks.copy()

In [72]:
# 'explodel 'leads' column to get a list of top actors 
explode_leads = one_df.explode('leads')


In [73]:
# convert 'director' column to string, (take out of list)
one_df['director'] = one_df['director'].str.get(0)

In [74]:
# create lists for the top actors and directors mentioned 

top_leads = [x for x in explode_leads.leads.value_counts().sort_values(ascending=False).head(100).index]
top_direct = [x for x in one_df.director.value_counts().sort_values(ascending=False).head(100).index]

In [75]:
# make new one-hot-encoded dataframe that will have 100 columns for the top 100 directors 

one_hot_director = one_df.copy()
for i in top_direct:
    one_hot_director[i] = np.where(one_hot_director['director']== i,1,0)
    

In [76]:
# drop everything aside from 100 actor columns and movie id

one_hot_director.drop(['leads', 'director', 'title'],axis=1,inplace=True)
one_hot_director.head(3)

id  John Ford  Michael Curtiz  Werner Herzog  Alfred Hitchcock  \
0    862          0               0              0                 0   
1   8844          0               0              0                 0   
2  15602          0               0              0                 0   

   Georges Méliès  Woody Allen  Sidney Lumet  Jean-Luc Godard  \
0               0            0             0                0   
1               0            0             0                0   
2               0            0             0                0   

   Charlie Chaplin  ...  Sidney J. Furie  Antonio Margheriti  Mario Bava  \
0                0  ...                0                   0           0   
1                0  ...                0                   0           0   
2                0  ...                0                   0           0   

   Bruce Beresford  Frank Borzage  Ken Loach  David Butler  James Ivory  \
0                0              0          0             0            0   
1                0              0          0             0            0   
2                0              0          0             0            0   

   Gerald Thomas  Jean Renoir  
0              0            0  
1              0            0  
2              0            0  

[3 rows x 101 columns]

Now let's one-hot the 'leads' feature 

In [77]:
# reduce features to just include the 'leads' we're encoding 
one_df = one_df[['id', 'title', 'leads']] 

In [78]:
# use list comprehension in for loop to one-hot encode 

one_hot_actor = one_df.copy()
for i in top_leads:
    one_hot_actor[i] = [1 if i in lst else 0 for lst in one_hot_actor.leads] 

In [79]:
# drop 'leads' column 

one_hot_actor.drop(['leads'],axis=1,inplace=True)

In [ ]:
# check out dataframe 

one_hot_actor[one_hot_actor['Tom Hanks'] == 1]

Pt. 6 Encoding the 'key words' column 

In [81]:
# create data frame with just keywords
movie_tags = movie_num[['title', 'id', 'keywords']]

In [82]:
# examine keyword value counts 
to_1D(movie_tags['keywords']).value_counts().head(10)


woman director      3011
independent film    1894
murder              1279
based on novel       815
musical              718
sex                  677
violence             646
nudity               628
revenge              616
biography            608
dtype: int64

In [83]:
# use explode function again to break out the values of 'keywords' column into new rows
explode_tags = movie_tags.explode('keywords')

In [84]:
# create list of top 1000 tags
top_tags = [x for x in explode_tags.keywords.value_counts().sort_values(ascending=False).head(100).index]

In [85]:
# create new data frame of tags one-hot-encoded 

one_hot_tags = movie_tags.copy()
for i in top_tags:
    one_hot_tags[i] = [1 if i in lst else 0 for lst in movie_tags.keywords] 

In [86]:
# drop 'keywords' column 

one_hot_tags.drop(['keywords'], axis=1,inplace=True)

In [87]:
# check out new df 
one_hot_tags.head(2)

title    id  woman director  independent film  murder  based on novel  \
0  Toy Story   862               0                 0       0               0   
1    Jumanji  8844               0                 0       0               0   

   musical  sex  violence  nudity  ...  slasher  wife husband relationship  \
0        0    0         0       0  ...        0                          0   
1        0    0         0       0  ...        0                          0   

   lawyer  supernatural  friends  dark comedy  college  student  comedy  fight  
0       0             0        1            0        0        0       0      0  
1       0             0        0            0        0        0       0      0  

[2 rows x 102 columns]

Pt. 7 Examine the features that are natively numerical 

In [88]:
movie_ints = movie_num[['id', 'title', 'release year', 'runtime', 'budget', 'revenue', 'vote_average', 'vote_count', 'popularity']]

In [89]:
# check if any values are null

movie_ints.runtime.isnull().value_counts()


False    44733
True       251
Name: runtime, dtype: int64

The 'runtime' column is the only one with null values(251). Let's see how many have '0' as the value

In [90]:
movie_ints[movie_ints['runtime'] == 0.0]

id                                  title  release year  runtime  \
222     61813                              Dream Man          1995      0.0   
224     62488             Destiny Turns on the Radio          1995      0.0   
398    172923                           Dos Crímenes          1995      0.0   
554    218473              The Beans of Egypt, Maine          1994      0.0   
667    221917                 The Run of the Country          1995      0.0   
...       ...                                    ...           ...      ...   
46528  374764                   How Most Things Work          2015      0.0   
46529  460135  LEGO DC Super Hero Girls: Brain Drain          2017      0.0   
46557  280422                            All at Once          2014      0.0   
46574  282308                 Whiffles, Cubic Artist          1912      0.0   
46588  222872                           Starquest II          1996      0.0   

       budget  revenue  vote_average  vote_count  popularity  
222         0        0           2.5           1    0.155859  
224         0        0           5.3           9    0.710671  
398         0        0           5.0           1    0.233376  
554         0        0           0.0           1    0.382470  
667         0        0           8.0           2    0.117662  
...       ...      ...           ...         ...         ...  
46528       0        0           6.8           2    0.274793  
46529       0        0          10.0           2    8.413734  
46557  750000        3           6.0           4    0.201582  
46574       0        0           0.0           0    0.003732  
46588       0        0           2.0           1    1.188052  

[1495 rows x 9 columns]

1,495 movies have zero for the runtime. Let's convert the '0' values to 'null' and impute them. 

In [91]:
# impute 'null' values 

movie_ints['runtime'] =  movie_ints['runtime'].replace(0, np.nan)

In [92]:
#impute missing runtimes with the average movie runtime 

movie_ints['runtime'].fillna(movie_ints['runtime'].mean(), inplace = True)

Let's take a look at the 'year' column. 

In [93]:
# create df with just release year

release_year = movie_ints[['id', 'release year']]

#drop 'release year' from num movie_ints

movie_ints.drop('release year', axis=1, inplace=True)

In [94]:
# Examining how many films are dated from the early 20th century or late 19th century 
  
early_century = release_year[(release_year['release year'] < 1930) & (release_year['release year'] > 0)]
early_century

id  release year
944     85638          1926
1321      653          1922
1836    28966          1927
1837    65203          1929
1921       19          1927
...       ...           ...
46605   44324          1904
46606  122036          1904
46608   49280          1900
46609  106807          1900
46622  227506          1917

[768 rows x 2 columns]

There are 80 films from before the year 1900. 71 films in the first decade of the 1900's. 169 films from the 1910's. 398 films from tge 1920's. 1228 films from the 30's. 

I am going to group all films from before 1930 into one date range (a total of 768 films). From 1930 onward, go by decades.

In [95]:
# How many films have no info for release year? 

release_year[release_year['release year'] == 0]

id  release year
718    365371             0
741    215107             0
3655   207731             0
5948    99885             0
13990   91205             0
...       ...           ...
46307  438910             0
46362  433711             0
46496  335251             0
46568  449131             0
46619  439050             0

[78 rows x 2 columns]

78 films are missing the release year info. Lets impute these '0' values

In [96]:
from sklearn.impute import IterativeImputer

In [97]:
# convert '0' to NaN value 

year_impute = release_year[['id', 'release year']]

year_impute['release year'] = year_impute['release year'].replace(0, np.nan)

# year_impute = year_impute.loc[710:750]
year_impute

id  release year
0         862        1995.0
1        8844        1995.0
2       15602        1995.0
3       31357        1995.0
4       11862        1995.0
...       ...           ...
46619  439050           NaN
46620  111109        2011.0
46621   67758        2003.0
46622  227506        1917.0
46623  461257        2017.0

[44984 rows x 2 columns]

In [98]:
# view list of "NaN" for release year
year_impute[year_impute['release year'].isna()]

id  release year
718    365371           NaN
741    215107           NaN
3655   207731           NaN
5948    99885           NaN
13990   91205           NaN
...       ...           ...
46307  438910           NaN
46362  433711           NaN
46496  335251           NaN
46568  449131           NaN
46619  439050           NaN

[78 rows x 2 columns]

We are going to use fancy imputer's 'Iterative Imputer' feature to fill in missing year values

In [99]:
imputer = fancy.IterativeImputer() # IterativeImputer is ok and KNN 
df = imputer.fit_transform(year_impute)

In [100]:
# convert our imputed array back into a data frame
year_impute = pd.DataFrame(df)

# name columns 
year_impute.columns = ['id', 'release year']

#make sure values are 'int' type 
year_impute[['id', 'release year']] = year_impute[[ 'id', 'release year']].astype(int)


Now lets join the imputed years df back with out release years df

In [101]:
release_year = year_impute.copy()


In [102]:
# convert 'release year' to string datatype

release_year['release year'] = release_year['release year'].astype(str)

In [103]:
#create new decade column based on films release year 

conditons = [release_year['release year'].str.contains('201'), 
            release_year['release year'].str.contains('200'),
            release_year['release year'].str.contains('199'),
            release_year['release year'].str.contains('198'),
            release_year['release year'].str.contains('197'),
            release_year['release year'].str.contains('196'),
            release_year['release year'].str.contains('195'),
            release_year['release year'].str.contains('194'),
            release_year['release year'].str.contains('193')]

choices = ['2010s', '2000s', '1990s', '1980s', '1970s', '1960s', '1950s', '1940s', '1930s']

release_year['decade'] = np.select(conditons, choices, default='Pre_1930')

In [104]:
# one hot encode the decade column 

decade_dummies = pd.get_dummies(release_year['decade'])
release_year = release_year.join(decade_dummies)

In [105]:
# drop 'decade' column 

release_year.drop(['decade'], axis=1,inplace=True)

In [106]:
# view our final 'release year' df with years binned by decade
release_year.head(5)

id release year  1930s  1940s  1950s  1960s  1970s  1980s  1990s  2000s  \
0    862         1995      0      0      0      0      0      0      1      0   
1   8844         1995      0      0      0      0      0      0      1      0   
2  15602         1995      0      0      0      0      0      0      1      0   
3  31357         1995      0      0      0      0      0      0      1      0   
4  11862         1995      0      0      0      0      0      0      1      0   

   2010s  Pre_1930  
0      0         0  
1      0         0  
2      0         0  
3      0         0  
4      0         0

Let's take a look at the dataframes we've created 

In [ ]:
movie_nati # df with nationality/language info
movie_comp # df with production companies 
movie_genres # df with genres
one_hot_director # top 1000 directors encoded
one_hot_actor # top 1000 actors encoded
one_hot_tags # top 1000 keywords encoded 
release_year # decade when the movie was released
movie_ints # native numerical stats for each movie

Now let's save each of these to a .csv file 

In [108]:
# save datasets

movie_nati.to_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Saved Datasets/movie_nat_lang.csv')
movie_comp.to_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Saved Datasets/movie_prodcomp.csv') 
movie_genres.to_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Saved Datasets/movie_genres.csv') 
one_hot_director.to_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Saved Datasets/movie_directors.csv') 
one_hot_actor.to_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Saved Datasets/movie_actors.csv') 
one_hot_tags.to_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Saved Datasets/movie_tags.csv') 
release_year.to_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Saved Datasets/release_year.csv')
movie_ints.to_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Saved Datasets/movie_ints.csv')

Pt. 8 Moving on to the Text Dataset

In [109]:
movie_text.head(2)

id      title                        genres  \
0   862  Toy Story   [Animation, Comedy, Family]   
1  8844    Jumanji  [Adventure, Fantasy, Family]   

                                            keywords  \
0  [jealousy, toy, boy, friendship, friends, riva...   
1  [board game, disappearance, based on children'...   

                                                cast         director  \
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  [John Lasseter]   
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst,...   [Joe Johnston]   

                                            overview  popularity  runtime  \
0  Led by Woody, Andy's toys live happily in his ...   21.946943     81.0   
1  When siblings Judy and Peter discover an encha...   17.015539    104.0   

     budget    revenue original_language     spoken_languages  \
0  30000000  373554033                en            [English]   
1  65000000  262797249                en  [English, Français]   

                                production_companies  \
0                          [Pixar Animation Studios]   
1  [TriStar Pictures, Teitler Film, Interscope Co...   

         production_countries  vote_average  vote_count  release year  
0  [United States of America]           7.7        5415          1995  
1  [United States of America]           6.9        2413          1995

1. Lets grab only the features that will be text-vectorized

In [110]:
movie_cats = movie_text[['id', 'title', 'genres', 'keywords', 'cast', 'director', 'production_companies']]

In [111]:
movie_cats

id                        title                        genres  \
0         862                    Toy Story   [Animation, Comedy, Family]   
1        8844                      Jumanji  [Adventure, Fantasy, Family]   
2       15602             Grumpier Old Men             [Romance, Comedy]   
3       31357            Waiting to Exhale      [Comedy, Drama, Romance]   
4       11862  Father of the Bride Part II                      [Comedy]   
...       ...                          ...                           ...   
46619  439050                       Subdue               [Drama, Family]   
46620  111109          Century of Birthing                       [Drama]   
46621   67758                     Betrayal     [Action, Drama, Thriller]   
46622  227506             Satan Triumphant                            []   
46623  461257                     Queerama                            []   

                                                keywords  \
0      [jealousy, toy, boy, friendship, friends, riva...   
1      [board game, disappearance, based on children'...   
2      [fishing, best friend, duringcreditsstinger, o...   
3      [based on novel, interracial relationship, sin...   
4      [baby, midlife crisis, confidence, aging, daug...   
...                                                  ...   
46619                                      [tragic love]   
46620                              [artist, play, pinoy]   
46621                                                 []   
46622                                                 []   
46623                                                 []   

                                                    cast            director  \
0      [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...     [John Lasseter]   
1      [Robin Williams, Jonathan Hyde, Kirsten Dunst,...      [Joe Johnston]   
2      [Walter Matthau, Jack Lemmon, Ann-Margret, Sop...     [Howard Deutch]   
3      [Whitney Houston, Angela Bassett, Loretta Devi...   [Forest Whitaker]   
4      [Steve Martin, Diane Keaton, Martin Short, Kim...     [Charles Shyer]   
...                                                  ...                 ...   
46619        [Leila Hatami, Kourosh Tahami, Elham Korda]  [Hamid Nematollah]   
46620  [Angel Aquino, Perry Dizon, Hazel Orencio, Joe...          [Lav Diaz]   
46621  [Erika Eleniak, Adam Baldwin, Julie du Page, J...    [Mark L. Lester]   
46622  [Iwan Mosschuchin, Nathalie Lissenko, Pavel Pa...  [Yakov Protazanov]   
46623                                                 []     [Daisy Asquith]   

                                    production_companies  
0                              [Pixar Animation Studios]  
1      [TriStar Pictures, Teitler Film, Interscope Co...  
2                         [Warner Bros., Lancaster Gate]  
3               [Twentieth Century Fox Film Corporation]  
4           [Sandollar Productions, Touchstone Pictures]  
...                                                  ...  
46619                                                 []  
46620                                      [Sine Olivia]  
46621                          [American World Pictures]  
46622                                        [Yermoliev]  
46623                                                 []  

[44984 rows x 7 columns]

2. Time to process some specific columns

In [112]:
#select top 3 actors from 'cast' column 
movie_cats['cast'] = movie_cats['cast'].apply(lambda x: x[:3] if len(x) >=3 else x)

In [113]:
# get ride of spaces in the 'cast' column and convert each word to lower case
movie_cats['cast'] = movie_cats['cast'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [114]:
movie_cats['director'] = movie_cats['director'].apply(lambda x: x*3)

In [115]:
# get ride of spaces in the 'director' column and convert each word to lower case
movie_cats['director'] = movie_cats['director'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [116]:
# Strip spaces and make genre names lower case 
movie_cats['genres'] = movie_cats['genres'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

# same for 'production companies'
movie_cats['production_companies'] = movie_cats['production_companies'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

3. Our 'keywords' column has a lot of variability and needs some special processing 

In [117]:
# We will convert our 'keywords' column to a 'series' 
s = movie_cats.apply(lambda x: pd.Series(x['keywords']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'keyword'

In [118]:
#examine our keywords 
s = s.value_counts()
s[:10]

woman director      3011
independent film    1894
murder              1279
based on novel       815
musical              718
sex                  677
violence             646
nudity               628
revenge              616
biography            608
Name: keyword, dtype: int64

In [119]:
# key words that occur once have no use, so we will eliminate them from the list 
s = s[s >1]

In [120]:
# defining a function that puts the keywords into a list 
def filter_keywords(x):
    words = []
    for i in x:
        if i in s:
            words.append(i)
    return words

In [121]:
from nltk.stem.snowball import SnowballStemmer
stemmer = SnowballStemmer('english')
stemmer.stem('dogs')

'dog'

In [122]:
# apply function to convert 'keywords' column to a list of words
movie_cats['keywords'] = movie_cats['keywords'].apply(filter_keywords)
# Apply stemmer to get the word stem
movie_cats['keywords'] = movie_cats['keywords'].apply(lambda x: [stemmer.stem(i) for i in x])
# Strip spaces and make keywords lower case 
movie_cats['keywords'] = movie_cats['keywords'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

4. Now that we've processed relevant columns we will merge them all into one column!

In [123]:
# create one column that is a combo of our other processed columns 
movie_cats['text_info'] = movie_cats['keywords'] + movie_cats['cast'] + movie_cats['director'] + movie_cats['genres'] + movie_cats['production_companies']
movie_cats['text_info'] = movie_cats['text_info'].apply(lambda x: ' '.join(x))

In [124]:
# drop processd columns
movie_cats.drop(['genres', 'keywords', 'cast', 'director', 'production_companies'], axis=1, inplace=True)

In [125]:
# take a look at out new condensed column 
movie_cats

id                        title  \
0         862                    Toy Story   
1        8844                      Jumanji   
2       15602             Grumpier Old Men   
3       31357            Waiting to Exhale   
4       11862  Father of the Bride Part II   
...       ...                          ...   
46619  439050                       Subdue   
46620  111109          Century of Birthing   
46621   67758                     Betrayal   
46622  227506             Satan Triumphant   
46623  461257                     Queerama   

                                               text_info  
0      jealousi toy boy friendship friend rivalri boy...  
1      boardgam disappear basedonchildren'sbook newho...  
2      fish bestfriend duringcreditssting oldmen walt...  
3      basedonnovel interracialrelationship singlemot...  
4      babi midlifecrisi confid age daughter motherda...  
...                                                  ...  
46619  tragiclov leilahatami kouroshtahami elhamkorda...  
46620  artist play pinoy angelaquino perrydizon hazel...  
46621  erikaeleniak adambaldwin juliedupage markl.les...  
46622  iwanmosschuchin nathalielissenko pavelpavlov y...  
46623             daisyasquith daisyasquith daisyasquith  

[44984 rows x 3 columns]

Our DataFrame looks good and ready for text vectorization!

In [126]:
# save dataset
movie_cats.to_csv('/Users/spence/Documents/GitHub/Movie Recommender System/Saved Datasets/movie_cats.csv')


Pt. 9 User Ratings

In [127]:
ratings

userId  movieId  rating   timestamp
0            1       31     2.5  1260759144
1            1     1029     3.0  1260759179
2            1     1061     3.0  1260759182
3            1     1129     2.0  1260759185
4            1     1172     4.0  1260759205
...        ...      ...     ...         ...
99999      671     6268     2.5  1065579370
100000     671     6269     4.0  1065149201
100001     671     6365     4.0  1070940363
100002     671     6385     2.5  1070979663
100003     671     6565     3.5  1074784724

[100004 rows x 4 columns]

In [128]:
#check for missing values
ratings.isnull().sum()

userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

There are no missing values in our ratings dataset. In the preproceesing/modeling step we will use 'collaborative filtering' to generate a recommendation for a user based on what they and similar users already like.

Looks like our 'wrangling' is done!!